### DISCUSSION TITLE
How do I tell it not to mess up my system?

### DISCUSSION THREAD

**Discussion by siddhsql** [Homebrew/discussions#5489] [Getting started]
### Output of `brew config`

```shell
❯ brew config                                                        ⏎
HOMEBREW_VERSION: 4.2.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: 0476c2e5e4f02b409084553896340b2f4ab3352e
Last commit: 3 months ago
Core tap JSON: 12 Apr 18:57 UTC
Core cask tap JSON: 12 Apr 18:57 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 14.0.3 build 1403
Git: 2.39.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5.2-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.1.


Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libcdt.5.dylib
  /usr/local/lib/libcgraph.6.dylib
  /usr/local/lib/libgvc.6.dylib
  /usr/local/lib/libgvpr.2.dylib
  /usr/local/lib/libpathplan.4.dylib
  /usr/local/lib/libxdot.4.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/graphviz/arith.h
  /usr/local/include/graphviz/cdt.h
  /usr/local/include/graphviz/cgraph.h
  /usr/local/include/graphviz/color.h
  /usr/local/include/graphviz/geom.h
  /usr/local/include/graphviz/graphviz_version.h
  /usr/local/include/graphviz/gvc.h
  /usr/local/include/graphviz/gvcext.h
  /usr/local/include/graphviz/gvcjob.h
  /usr/local/include/graphviz/gvcommon.h
  /usr/local/include/graphviz/gvconfig.h
  /usr/local/include/graphviz/gvplugin.h
  /usr/local/include/graphviz/gvplugin_device.h
  /usr/local/include/graphviz/gvplugin_layout.h
  /usr/local/include/graphviz/gvplugin_loadimage.h
  /usr/local/include/graphviz/gvplugin_render.h
  /usr/local/include/graphviz/gvplugin_textlayout.h
  /usr/local/include/graphviz/gvpr.h
  /usr/local/include/graphviz/pack.h
  /usr/local/include/graphviz/pathgeom.h
  /usr/local/include/graphviz/pathplan.h
  /usr/local/include/graphviz/textspan.h
  /usr/local/include/graphviz/types.h
  /usr/local/include/graphviz/usershape.h
  /usr/local/include/graphviz/xdot.h

Warning: Unbrewed '.la' files were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.la' files:
  /usr/local/lib/libcdt.la
  /usr/local/lib/libcgraph.la
  /usr/local/lib/libgvc.la
  /usr/local/lib/libgvpr.la
  /usr/local/lib/libpathplan.la
  /usr/local/lib/libxdot.la

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/libcdt.pc
  /usr/local/lib/pkgconfig/libcgraph.pc
  /usr/local/lib/pkgconfig/libgvc.pc
  /usr/local/lib/pkgconfig/libgvpr.pc
  /usr/local/lib/pkgconfig/libpathplan.pc
  /usr/local/lib/pkgconfig/libxdot.pc
```

### Description of issue

I will get some hate responses for this but every time I use this tool, it often messes up my system and programs that used to work stop working. e.g., see [this](https://github.com/orgs/python-poetry/discussions/9305). is there any way to tell it not to change the system settings so that other programs do not break? it seems to be doing lot of config changes underneath and shouldn't be installing extra baggage and doing unnecessary things.

**Comment by p-linnane** [comment]
There are a few things to start with:

1. Your macOS version is from September 2023. You should update it. If you don't want to update to macOS 14 for some reason, you should at least update to macOS 13.6.7, which came out in May 2024.
2. You haven't updated `brew` since April. You should update it with `brew update`.
3. Did you migrate from an Intel Mac at some point? Your brew doctor warnings are showing a bunch of things in `/usr/local` which is the Intel Homebrew prefix. You may want to clear those things out. On arm64 Macs, everything is in `/opt/homebrew`

Start with those 3 things and then see how it goes.

**Comment by fvsch** [comment]
I find that Homebrew doesn't mess up my system. Things might break, but the breakage is usually contained to:

1. software installed with Homebrew;
2. software installed in a different way and which ends up depending on software installed with Homebrew (if only because of their presence in the `$PATH`).

The issue you linked with `poetry` seems to be an example of the second type, if not the first one. None of those are “messing up my system” in my book. I think you should differentiate messing up your system and breaking your workflow for whatever tooling you are currently using:

- "mess up my system": your OS is not booting, or built-in software crashes.
- "mess up my workflow": some of the software you installed on top of your system is crashing, and you might need to revise your setup, which you _can_ do because your system is still working.

One issue I’ve faced sometimes is when installing a formula like `python` or `php` with Homebrew, because those formulae can update to a new minor version, and minor versions of programming languages often come with a brand new set of configuration files and globally installed packages (when installing any package globally). To avoid this issue, you can prefer installing a specific minor version (e.g. with `brew install python@3.12`), and then manually add its `bin` directory to your `PATH`.

Though for programming language runtimes, I prefer not using Homebrew. Especially because they tend to come with a way to globally install packages (`gem`, `pip`, `npm -g`, etc.); while those usually get installed in a folder managed by Homebrew or in a hidden folder in your `$HOME`, and can be safely removed (possibly breaking your _workflow_ but not your _system_). If possible, I try to use more explicitly self-contained ways to install programming language runtimes and packages, such as:

- Doing everything for a project in a container (using Docker, Podman, DevContainers, etc.).
- Or using a tool version manager like [Mise](https://mise.jdx.dev/).

Looking at the [Poetry installation docs](https://python-poetry.org/docs/#installation), they also have a very big warning about installing Poetry in a virtual environment to avoid the very issue you are reporting:

> Poetry should always be installed in a dedicated virtual environment to isolate it from the rest of your system. It should in no case be installed in the environment of the project that is to be managed by Poetry. This ensures that Poetry’s own dependencies will not be accidentally upgraded or uninstalled.

So I would look at things like:

- What Python version am I using?
- How is it installed and updated? How can I pin it? (Using Homebrew or something else.)
- Can I reinstall Poetry cleanly with one of their recommended methods?
